### PR TITLE
Validator parser

### LINKED ISSUE #4: Implement validator module
`Validator` module represent the parsing of validator language:

```
Validator {
    Process_name: {
       repeat between 1 and n {

      },
      maybe {

      }
   }
}
```

### PR DESCRIPTION
**Pull request by Kn99HN**
### Description
- Parse validator into a recognizer struct with list of nested statements
- Providing unit tests
- Parsing "task", "maybe", "statement", "send", "receive"

The output has the form
```
%Validator.Recognizer{
   name: "A",
   map: { 
      "process_A" => [
              %Validator.Notice{ pattern: "C" }
          ]
   }
}
```

Each recognizer comprises of a map that maps from process names to its validator statements. Only `task`, `maybe` and `repeat` can have nested statements. Those are represented as a `[any()]` in `statements` key of each type of struct.
### Note for reviewer
- Sorry for such long PR : (
- Please run `mix test` and ensure that the tests are passing

### CODE REVIEW

**Review by joyceluyy — APPROVED** [review]

File: apps/validator/lib/validator.ex [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+    %Validator.Receive{
+      name: received_process
+    }
+  end
+
+  defp parse_task(s) do
+    task_name = get_pattern_from_expectation(s)
+
+    %Validator.Task{
```
Comment (Kn99HN) [inline apps/validator/lib/validator.ex:25]: Could be shortened to just use the function.

File: apps/validator/lib/validator.ex [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      name: validator_name
+    }
+  end
+
+  defp find_indexes(tokens, index, ls) do
+    if index >= length(tokens) do
+      nil
+    else
+      curr = Enum.at(tokens, index)
+      {char, idx} = Enum.at(ls, 0, {"", 0})
```
Comment (Kn99HN) [inline apps/validator/lib/validator.ex:57]: Could be changed to use `case ... do` here.

File: apps/validator/lib/validator.ex [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    recognizer: {
+      map: {
+        process_name => [{expectation}]
+      }
+    }
+  }
+  """
+  @spec parse_recognizer([any()], %Validator.Recognizer{}) :: %Validator.Recognizer{}
+  defp parse_recognizer(tokens, recognizer) do
+    if length(tokens) == 0 do
```
Comment (Kn99HN) [inline apps/validator/lib/validator.ex]: can be changed to `case ... do`

**Review by Kn99HN — COMMENTED** [review]
Minor nit picks to change.

**Review by joyceluyy — APPROVED** [review]

**Pull request merged**